### PR TITLE
feat: implement dedicated server data source

### PR DESCRIPTION
**Pull request by majidkarimizadeh**

### CODE REVIEW

File: .env.dist [hunk 1]
```diff
@@ -1,2 +1,3 @@
 PUBLIC_CLOUD_API_SPEC_URL=https://raw.githubusercontent.com/Leaseweb/leaseweb-go-sdk/main/publicCloud/api/openapi.yaml
+DEDICATED_SERVER_API_SPEC_URL=https://raw.githubusercontent.com/Leaseweb/leaseweb-go-sdk/main/dedicatedServer/api/openapi.yaml
 #PUBLIC_CLOUD_API_SPEC_URL=http://host.docker.internal:8081/publicCloud.json
```
Comment (alrayyes) [inline .env.dist:3]: The url for local development should be added here

`#DEDICATED_SERVER_API_SPEC_URL=http://host.docker.internal:8081/dedicatedServer.json`

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,5 @@
+docker-compose.override.yml
```
Comment (alrayyes) [inline .gitignore:1]: override doesn't exist (yet)

File: internal/core/domain/dedicated_server/dedicated_server.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package domain
+
+type DedicatedServer struct {
+	Id string
+}
+
+func NewDedicatedServer(id string) DedicatedServer {
+	dedicatedServer := DedicatedServer{Id: id}
```
Comment (alrayyes) [inline internal/core/domain/dedicated_server/dedicated_server.go:8]: Id could be a value object. We only allow strings comprised of integers here.

File: internal/core/domain/dedicated_server/dedicated_servers.go [hunk 4]
```diff
@@ -0,0 +1,3 @@
+package domain
+
+type DedicatedServers []DedicatedServer
```
Comment (alrayyes) [inline internal/core/domain/dedicated_server/dedicated_servers.go:3]: Nice!

File: internal/core/ports/dedicated_server_repository.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import (
+	"context"
+
+	domain "github.com/leaseweb/terraform-provider-leaseweb/internal/core/domain/dedicated_server"
+	"github.com/leaseweb/terraform-provider-leaseweb/internal/repositories/shared"
+)
+
+// DedicatedServerRepository is used to connect to dedicated_server api.
+type DedicatedServerRepository interface {
+	// GetAllDedicatedServers Retrieve all dedicated_servers from the dedicated server api.
```
Comment (alrayyes) [inline internal/core/ports/dedicated_server_repository.go:12]: This should be a proper sentence. So `GetAllDedicatedServers retrieves all dedicated_servers from the deciated server api.`.

This is how all the go documentation is written.

File: internal/core/services/dedicated_server/service.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/leaseweb/terraform-provider-leaseweb/internal/core/ports"
+	"github.com/leaseweb/terraform-provider-leaseweb/internal/core/services/errors"
+)
+
+// Service fulfills the contract for ports.DedicatedServerService.
+type Service struct {
+	dedicatedServerRepository ports.DedicatedServerRepository
+}
+
+func (srv Service) GetAllDedicatedServers(ctx context.Context) (
```
Comment (alrayyes) [inline internal/core/services/dedicated_server/service.go:16]: I don't know what the best practice is but I like to return nil if an error is found. `GetAllDedicatedServers(ctx context.Context) (*domain.DedicatedServers, *errors.ServiceError)`

This way it's clear that dedicatedServers is not set.

Comment (alrayyes) [reply internal/core/services/dedicated_server/service.go:16]: Then you can just `return nil, *err`

File: internal/facades/dedicated_server/data_adapters/to_data_source_model/adapter.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	domain "github.com/leaseweb/terraform-provider-leaseweb/internal/core/domain/dedicated_server"
+	"github.com/leaseweb/terraform-provider-leaseweb/internal/provider/data_sources/dedicated_server/model"
+)
+
+func AdaptDedicatedServers(domainDedicatedServers domain.DedicatedServers) model.DedicatedServers {
+	var dedicatedServers model.DedicatedServers
+
+	for _, domainDedicatedServer := range domainDedicatedServers {
+		dedicated_server := adaptDedicatedServer(domainDedicatedServer)
```
Comment (alrayyes) [inline internal/facades/dedicated_server/data_adapters/to_data_source_model/adapter.go:13]: I don't know what best practice is in golang, this or `dedicatedServer` for variable names.

**Review by alrayyes — CHANGES_REQUESTED** [review]
Looks good, a few small things & missing acceptance tests.

File: internal/facades/dedicated_server/data_adapters/to_data_source_model/adapter.go (hunk 7 above)
Comment (majidkarimizadeh) [reply internal/facades/dedicated_server/data_adapters/to_data_source_model/adapter.go:13]: missed this one, will change it to `dedicatedServer`

File: .gitignore (hunk 2 above)
Comment (majidkarimizadeh) [reply .gitignore:1]: it does exist on my system, so just ignored it

File: internal/provider/data_sources/dedicated_server/dedicated_servers/read.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+			err.ErrorResponse,
+			&resp.Diagnostics,
+			"Unable to read dedicated servers",
+			err.Error(),
+		)
+
+		return
+	}
+
+	state := dedicatedSeervers
```
Comment (imtiazPabel) [inline internal/provider/data_sources/dedicated_server/dedicated_servers/read.go:33]: I believe that's typo

File: internal/repositories/dedicated_server_repository/dedicated_server_repository.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+	result, response, err := request.Execute()
+
+	if err != nil {
+		return nil, shared.NewSdkError("GetAllDedicatedServers", err, response)
+	}
+
+	metadata := result.GetMetadata()
+	pagination := shared.NewPagination(
+		50, // metadata.GetLimit(),
```
Comment (imtiazPabel) [inline internal/repositories/dedicated_server_repository/dedicated_server_repository.go:55]: Curious minds want to know, any specific reason to set the value 50?

Comment (majidkarimizadeh) [reply internal/repositories/dedicated_server_repository/dedicated_server_repository.go:55]: with the default limit we face `timeout`, that's why I hard coded to 50, we need to remove the hard coded part later on.

File: internal/core/ports/dedicated_server_repository.go (hunk 5 above)
Comment (alrayyes) [reply internal/core/ports/dedicated_server_repository.go:12]: typo: Retrieve := retrieves

**Review by alrayyes — APPROVED** [review]

**Pull request merged**